packages/widgets/user-profile-widget/src/lib/widget/mixins/initMixin/initComponentsMixins/initUserCustomAttributesMixin.ts

```diff
@@ -0,0 +1,41 @@
+import { UserAttributeDriver } from '@descope/sdk-component-drivers';
+import { compose, createSingletonMixin } from '@descope/sdk-helpers';
+import { loggerMixin } from '@descope/sdk-mixins';
+import { getUserCustomAttrs } from '../../../state/selectors';
+import { stateManagementMixin } from '../../stateManagementMixin';
+import { initWidgetRootMixin } from './initWidgetRootMixin';
+
+export const initUserCustomAttributesMixin = createSingletonMixin(
+  <T extends CustomElementConstructor>(superclass: T) =>
+    class UserCustomAttributesMixinClass extends compose(
+      stateManagementMixin,
+      loggerMixin,
+      initWidgetRootMixin,
+    )(superclass) {
+      customValueUserAttr: UserAttributeDriver;
+
+      #initCustomValueUserAttrs() {
+        const allCustomAttributesComponents = this.shadowRoot?.querySelectorAll(
+          'descope-user-attribute[data-id^="customAttributes."]',
+        );
+
+        Array.from(allCustomAttributesComponents).forEach((nodeEle) => {
+          const attrName = nodeEle.getAttribute('data-id');
+          const customAttrName = attrName.replace('customAttributes.', '');
+
+          const compInstance = new UserAttributeDriver(nodeEle, {
+            logger: this.logger,
+          });
+
+          const userCustomAttributesData = getUserCustomAttrs(this.state);
```

Comment:
There is no need to run the selector for every component, we should run it once

In addition, we should register to state changes, so whenever we will get a new state, in case the custom attributes are changed, the components will update

Reply:
good call about this line, changed it's scope.

Since custom attributes are in the stateManagementMixin I believe it is being observed. 
Correct me if I am wrong.

![Uploading image.png…]()